feat: Make stream-only select patterns behave like `stream.*`

tayloramurphy opened the pull request:
Closes https://github.com/meltano/meltano/issues/9427 https://github.com/meltano/meltano/issues/3024 

## Summary

Updates the `select` function to automatically select the stream when only stream.property patterns are present (e.g., `"users.name"`) 

## Changes

### Testing
- Added comprehensive test coverage verifying the new behavior
- Tests confirm CLI integration works correctly with the enhanced patterns
- Validated that `select_filter` continues to operate independently at the stream level

### Documentation
- Updated `select` extra documentation to explain property-level selection patterns
- Completely rewrote `select_filter` documentation with comparison table and use cases
- Enhanced CLI documentation to clarify the distinction between configuration-time property selection and runtime stream filtering
- Added practical examples demonstrating the two-phase selection architecture

## Testing

Run the test suite with focus on catalog selection:
```bash
pytest tests/meltano/core/plugin/singer/test_catalog.py -v
```

## Summary by Sourcery

Implement automatic stream selection for property-level patterns in the `select` extra, expand test coverage on selection and filtering rules, and update documentation and CLI references to clarify the interaction between `select` and `select_filter`

New Features:
- Automatically select the stream when only a stream.property pattern is used in the `select` extra
- Differentiate and document the two-phase selection architecture (`select` vs `select_filter`) with practical examples

Enhancements:
- Rewrite and expand documentation for `select` and `select_filter` extras, including comparison tables and CLI examples
- Improve CLI help text to clarify stream-level filtering and property-level selection distinctions

Documentation:
- Update conceptual and CLI reference docs to reflect automatic stream selection and two-phase selection workflow

Tests:
- Add tests for `select_metadata_rules` to verify stream and nested property rule generation
- Add tests for `select_filter_metadata_rules` to verify CLI integration and behavior differences

Commit:
feat: Update select function to treat stream-only patterns as stream.*

When a user specifies just a stream name (e.g., "users"), it now behaves
identically to "users.*", selecting the entire stream and all properties.
Previously, stream-only patterns only created a stream-level selection rule,
while "stream.*" patterns created both stream and property selection rules.

This change ensures consistent behavior:
- "users" and "users.*" now generate identical metadata rules
- Both create stream selection + wildcard property selection rules
- Stream.property patterns continue to work as before

The change maintains backward compatibility while clarifying the semantic
intent of stream-only selections.

Commit:
test: add comprehensive tests for select and select_filter behavior

- Test stream-only patterns in select behave like stream.*
- Test stream.property patterns create both stream and property rules
- Test CLI --select/--exclude integration with stream-only patterns
- Test behavioral differences between select and select_filter functions
- Test that select_filter preserves existing property selections
- All 29 tests pass (24 existing + 5 new)

Commit:
docs: Clarify select vs select_filter behavior and usage

- Document that select operates at property level, select_filter at stream level
- Explain the new stream-only pattern behavior ("users" equals "users.*")
- Add comparison table and practical examples showing how both work together
- Update CLI docs to clarify --select/--exclude preserve property selections
- Improve examples throughout to demonstrate real-world usage patterns

Commit:
test: Add test for nested property selection behavior

Verifies that nested property selections (e.g., users.address.city) continue
to work correctly and are not affected by the stream-only pattern enhancement.
The test confirms that patterns with explicit property paths do not create
wildcard rules and maintain their specific selection behavior.

Commit:
Update CLI example to show deeply nested properties

tayloramurphy commented:
I don't think this PR fully addresses https://github.com/meltano/meltano/issues/9422 For that issue I think a mermaid style visual would actually be helpful.

visch commented on docs/docs/concepts/plugins.mdx:
```diff
@@ -517,19 +535,35 @@ export TAP_SMOKE_TEST__SELECT='["animals\\.data.id","animals\\.data.verified"]'
-Selection filter rules use entity identifiers that correspond to Singer stream `tap_stream_id` values. Rules indicating that an entity should be excluded are prefixed with an exclamation mark (`!`). [Unix shell-style wildcards](<https://en.wikipedia.org/wiki/Glob_(programming)#Syntax>) can be used in entity identifiers to match multiple entities at once.
+| `select` | `select_filter` |
+|----------|----------------|
+| **Property-level**: Controls which properties within streams are included | **Stream-level**: Controls which entire streams are included/excluded |
```
Isn't this even less true with this PR? Now stream only select patterns don't "Controls which properties within streams are included"

tayloramurphy replied:
UPdated to clarify

visch commented on docs/docs/concepts/plugins.mdx:
```diff
@@ -517,19 +535,35 @@ export TAP_SMOKE_TEST__SELECT='["animals\\.data.id","animals\\.data.verified"]'
+| `select` | `select_filter` |
+|----------|----------------|
+| **Property-level**: Controls which properties within streams are included | **Stream-level**: Controls which entire streams are included/excluded |
+| **Configuration-time**: Set during project configuration | **Runtime**: Typically set via CLI options (`--select`, `--exclude`) |
```
? Runtime I would say would typically be set via ENV vars, not CLI options as [run ](https://deploy-preview-9434--meltano.netlify.app/reference/command-line-interface/#parameters-3)doesn't have a parameter for select like those listed 

Very confusing comparison to me

visch commented on docs/docs/concepts/plugins.mdx:
```diff
@@ -517,19 +535,35 @@ export TAP_SMOKE_TEST__SELECT='["animals\\.data.id","animals\\.data.verified"]'
+|----------|----------------|
+| **Property-level**: Controls which properties within streams are included | **Stream-level**: Controls which entire streams are included/excluded |
+| **Configuration-time**: Set during project configuration | **Runtime**: Typically set via CLI options (`--select`, `--exclude`) |
+| **Purpose**: Define what data fields you want | **Purpose**: Filter which streams to process in a specific run |
```
+1

visch commented on docs/docs/concepts/plugins.mdx:
```diff
@@ -517,19 +535,35 @@ export TAP_SMOKE_TEST__SELECT='["animals\\.data.id","animals\\.data.verified"]'
+| **Property-level**: Controls which properties within streams are included | **Stream-level**: Controls which entire streams are included/excluded |
+| **Configuration-time**: Set during project configuration | **Runtime**: Typically set via CLI options (`--select`, `--exclude`) |
+| **Purpose**: Define what data fields you want | **Purpose**: Filter which streams to process in a specific run |
+| **Preserves existing selections**: No, defines the selection | **Preserves existing selections**: Yes, filters on top of existing `select` configuration |
```
+1

visch commented on docs/docs/concepts/plugins.mdx:
```diff
@@ -517,19 +535,35 @@ export TAP_SMOKE_TEST__SELECT='["animals\\.data.id","animals\\.data.verified"]'
-Entity names can be discovered using [`meltano select --list --all <plugin>`](/reference/command-line-interface#select).
+## When to Use `select_filter`
+
+- **Runtime filtering**: When you want to process only certain streams for a specific pipeline run
```
via Env vars is what you're trying to say right? Seems to me like this isn't a "When" it's a "How"

tayloramurphy replied:
removed

visch commented on docs/docs/concepts/plugins.mdx:
```diff
@@ -517,19 +535,35 @@ export TAP_SMOKE_TEST__SELECT='["animals\\.data.id","animals\\.data.verified"]'
+- **Runtime filtering**: When you want to process only certain streams for a specific pipeline run
+- **Testing/debugging**: Temporarily include/exclude streams without changing your base configuration
+- **Conditional processing**: Different runs need different subsets of your configured streams
+- **CLI operations**: Via [`meltano el --select`](#el) or [`meltano el --exclude`](#el)
```
Run doesn't have this though right? Should this be in the "runtime filtering" bullet point above?

tayloramurphy replied:
correct

visch commented on docs/docs/concepts/plugins.mdx:
```diff
@@ -517,19 +535,35 @@ export TAP_SMOKE_TEST__SELECT='["animals\\.data.id","animals\\.data.verified"]'
+| **Preserves existing selections**: No, defines the selection | **Preserves existing selections**: Yes, filters on top of existing `select` configuration |
 
-Entity names can be discovered using [`meltano select --list --all <plugin>`](/reference/command-line-interface#select).
+## When to Use `select_filter`
```
Parallelization is the best use case I've had, ie "Parallelization: Run an instance of meltano to sync a single stream"

visch commented on docs/docs/concepts/plugins.mdx:
```diff
@@ -423,9 +423,15 @@ that are applied to the extractor's [discovered catalog file](https://hub.meltan
 
-Entity and attribute names can be discovered using [`meltano select --list --all <plugin>`](/reference/command-line-interface#select).
+A selection rule can be either:
+- **Stream-only pattern** (e.g., `"users"`): Selects the stream and all its properties, equivalent to `"users.*"`
```
Before didn't this select the primary key(s)? I could be mistaken but if so it'd be a breaking change for some folks? Maybe not a huge deal

tayloramurphy replied:
@visch this still does select the primary keys, but yes it would select more by default now. I'd argue that's the more intuitive and expected behavior. If I select a table I'd likely be surprised just to see an ID and timestamp flow over. 

Not sure if this would warrant a full new v4 version, but I'd defer to @edgarrmondragon on that.

edgarrmondragon replied:
I think this is the thing that made me reconsider https://github.com/meltano/meltano/pull/8810 and why it's still a draft.

It does the `inclusion: automatic` Singer metadata[^1] a bit redundant, unless I'm missing something.

[^1]: https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#metadata

visch reviewed (changes requested):
Definitely better than before but it's hard for me to think about the doc improvement and the behavior change at the same time 🤷

Commit:
Incorporate PR feedback

Commit:
fix: Remove stream-only wildcard selection behavior

Previously, selecting a stream-only pattern like 'users' would behave
the same as 'users.*', selecting all properties. This has been removed.

Current behavior:
- Stream-only patterns (e.g., 'users') now only select the stream itself
- Stream.property patterns (e.g., 'users.name') select both the stream
  AND the specific property (preserving existing behavior)
- Stream.* patterns continue to select the stream and all properties

This change makes the selection behavior more explicit and predictable,
requiring users to use 'stream.*' when they want all properties selected.

Commit:
Update docs to reflect the correct behavior.

tayloramurphy commented:
@sourcery-ai review

Commit:
Update src/meltano/core/plugin/singer/catalog.py

Co-authored-by: sourcery-ai[bot] <58596630+sourcery-ai[bot]@users.noreply.github.com>

Commit:
Update docs/docs/reference/command-line-interface.md

Co-authored-by: sourcery-ai[bot] <58596630+sourcery-ai[bot]@users.noreply.github.com>

Commit:
Merge branch 'main' into update-select-extra-behavior

edgarrmondragon committed:
Merge branch 'main' into update-select-extra-behavior

edgarrmondragon committed:
Merge branch 'main' into update-select-extra-behavior

edgarrmondragon committed:
Merge branch 'main' into update-select-extra-behavior

edgarrmondragon committed:
Merge branch 'main' into update-select-extra-behavior

edgarrmondragon committed:
Merge branch 'main' into update-select-extra-behavior

edgarrmondragon committed:
Merge branch 'main' into update-select-extra-behavior

edgarrmondragon committed:
Use constant

edgarrmondragon commented on tests/meltano/core/plugin/singer/test_catalog.py:
```diff
@@ -1098,6 +1098,166 @@ def test_node_selection(self, node: dict, selection_type: SelectionType) -> None
         """Test that selection metadata produces the expected selection type member."""
         assert ListSelectedExecutor.node_selection(node) == selection_type
 
+    def test_select_metadata_rules_stream_property_creates_both_rules(self) -> None:
```
These tests pass if I restore `catalog.py` to the `main` version, so I think it's worth reverting 33bc2017b867ef3020272ec9a4f5d78650c64b50 now that we want to actually target v4.

edgarrmondragon committed:
Revert "fix: Remove stream-only wildcard selection behavior"

This reverts commit 33bc2017b867ef3020272ec9a4f5d78650c64b50.

edgarrmondragon committed:
Revert "Update docs to reflect the correct behavior."

This reverts commit 6c25b6897012c8f948dc440fcc5333c03157fa75.

edgarrmondragon committed:
Add new test case

edgarrmondragon commented:
I also think this is complimentary to https://github.com/meltano/meltano/pull/8810.

edgarrmondragon committed:
Update else condition
